### PR TITLE
feat: add launch token claiming process

### PR DESCRIPTION
**Pull request by wischli**
## fixes KILTProtocol/ticket#1172
Adds `pallet_vesting` and custom `kilt_launch` pallet which enable
* Enable setting vesting and KILT custom balance lock in the Genesis config
* Add extrinsic for the claiming process of migrating tokens, vesting and locks from an unowned Genesis-block address to a user-owned account address
  * either from a single source or a vector of sources to a single target address
* Add extrinsic to change transfer account
* Add extrinsic to force_unlock
* Add extrinsic to transfer KILT custom locked tokens

## ToDo

- [x] Update unit tests to make use of `vest` which was made public in c54b7a4
- [x] Benchmarks

## How to test:

It can also be tested in the Polkadot Apps but I suppose it makes more sense if I walk you through this. 

In short: During the genesis block building, the transfer account and a bunch of pseudo accounts (whose addresses etc. are described in <you_know_where>) for vesting and locks are set up. For each of those pseudo accounts, you need to call `migrate_genesis_account` or `migrate_multiple_genesis_accounts` signed by the transfer account. Then, either vesting or the custom lock should be set up. 

```
cargo test --all --all-features
```

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

**Comment by wischli** [comment]
@weichweich I would still like to add benchmarks before we merge this but since this is the only piece left missing, you could already review it. Please be as rigorous as possible!

File: pallets/kilt-launch/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+				<BalanceLocks<T>>::remove(&source);
+
+				// Only needs to be handled in the case of a `locked_transfer`, e.g., when
+				// `max_amount` is set because else the source address is never added to
+				// `UnlockingAt`
+				if max_amount.is_some() {
+					let remove_source_map: Vec<T::AccountId> = <UnlockingAt<T>>::take(unlock_block)
+						.unwrap_or_default()
+						.into_iter()
+						.filter_map(|acc_id| if &acc_id == source { None } else { Some(acc_id) })
```
Comment (weichweich) [inline pallets/kilt-launch/src/lib.rs:585]:
```suggestion
						.filter(|acc_id| &acc_id == source)
```

**Review by weichweich — CHANGES_REQUESTED** [review]
Looks good! Nice code.  🙂 Few of minor style thinks. Probably very subjective. And some questions...

File: pallets/kilt-launch/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+		/// unlock balance from vesting.
+		///
+		/// Note: Setting the custom lock actually only occurs in this call (and
+		/// not when building the genesis block) to avoid overhead from handling
+		/// locks when migrating. We can do so because all target accounts
+		/// are not owned by anyone and thus these cannot sign and/or call any
+		/// extrinsics.
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(9, 7))]
+		#[transactional]
+		pub(super) fn migrate_genesis_account(
```
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs:292]: cannot be called from another pallet?

File: pallets/kilt-launch/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+			if let Some(lock) = locks
+				.iter()
+				.find(|BalanceLock::<<T as pallet_balances::Config>::Balance> { id, .. }| id == &KILT_LAUNCH_ID)
+			{
+				ensure!(lock.amount >= amount, Error::<T>::InsufficientLockedBalance);
+
+				// We can substract because of the above check
+				let amount_new = T::ExistentialDeposit::get().max(lock.amount - amount);
+
+				if amount_new <= T::ExistentialDeposit::get() {
```
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs:378]: and i would maybe still use a `saturating_sub` instead of the plain one. since the `-` panics

File: pallets/kilt-launch/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+	///
+	/// Required for the claiming process.
+	#[pallet::storage]
+	#[pallet::getter(fn get_transfer_account)]
+	pub type TransferAccount<T> = StorageValue<_, <T as frame_system::Config>::AccountId>;
+
+	/// Maps a block to account ids which have their balance locked.
+	///
+	/// Required for automatic unlocking once the block number is reached in
+	/// `on_initialize`.
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:199]: Yeah, that's one of the missing pieces.

File: pallets/kilt-launch/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+					);
+
+					// Add unlock block to storage
+					<BalanceLocks<T>>::insert(
+						who,
+						LockedBalance::<T> {
+							block: *length,
+							amount: *locked,
+						},
+					);
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:145]: Good point. Theoretically, we will go through the list of accounts million times by hand to be sure everything is correct but it will still be good to add a check here for sure.

File: pallets/kilt-launch/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+
+		// Transfer to target addess
+		let amount = <pallet_balances::Pallet<T>>::total_balance(source);
+		let _ = <pallet_balances::Pallet<T> as Currency<T::AccountId>>::transfer(source, target, amount, AllowDeath)?;
+
+		// Migrate vesting info and set the corresponding vesting lock if necessary
+		let mut post_weight: Weight = Self::migrate_vesting(source, target)?;
+
+		// Set the KILT custom lock if necessary
+		post_weight += Self::migrate_kilt_balance_lock(source, target, None)?;
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:452]: Wouldn't the transaction fail and since it's defined with the `transactional` feature, none of the changes would actually applied?

[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:452]: As a note: There will be no genesis address with vesting and the kilt lock, there will be at most one of these for each genesis address.

File: pallets/kilt-launch/src/lib.rs (hunk 1 above)
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:585]: 🙈  embarrassing

File: pallets/kilt-launch/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,608 @@
...
+		/// definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Maximum number of claims which can be migrated in a single call.
+		/// Used for weight estimation.
+
+		/// NOTE:
+		/// + Benchmarks will need to be re-run and weights adjusted if this
+		/// changes. + This pallet assumes that dependents keep to the limit
+		/// without enforcing it.
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:91]: This was actually copied 1:1 from the [collective pallet](https://github.com/paritytech/substrate/blob/787ea591eaf9cd2e8b0d638c87d9745762047aa3/frame/collective/src/lib.rs#L146) 🥲  

The second sentence says that there is a maximum limit, but we are not throwing if this limit is exceeded. This statement used to be true for our pallet, but now it's plain false.

File: pallets/kilt-launch/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,608 @@
...
+			ensure!(Some(who) == <TransferAccount<T>>::get(), Error::<T>::Unauthorized);
+
+			ensure!(sources.len() < T::MaxClaims::get(), Error::<T>::ExceedsMaxClaims);
+
+			// TODO: How to do this with map?
+			let mut post_weight: Weight = 0;
+			for s in sources.into_iter() {
+				let source = T::Lookup::lookup(s)?;
+				post_weight += Self::migrate_user(&source, &target)?;
+			}
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:332]: I actually like your approach more but have to agree, that mine is more readable. Should I roll a dice?

File: pallets/kilt-launch/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs:452]: Ah didn't see the `transactional` 😅

File: pallets/kilt-launch/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,608 @@
...
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+// TODO: Add benchmarking
+// #[cfg(feature = "runtime-benchmarks")]
+// mod benchmarking;
+
+pub const KILT_LAUNCH_ID: LockIdentifier = *b"kiltlock";
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:65]: I actually had `kiltlaun` before and then changed it because it's also displayed in the Polkadot Apps. But I agree with you.

File: pallets/kilt-launch/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs:332]: mine is one line less! :D Naaaah leave it as it is... 🤷🏻

File: pallets/kilt-launch/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:292]: `pub` seems to be sufficient for our use-case

File: pallets/kilt-launch/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,608 @@
...
+		type MaxClaims: Get<usize>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub balance_locks: Vec<(T::AccountId, T::BlockNumber, <T as pallet_balances::Config>::Balance)>,
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:101]: We eventually want it to be a map from `AccountId --> LockedBalance` but that happens in the genesis build for which we want a vector of accounts as input. It's done similarly in other pallets like [`vesting`](https://github.com/paritytech/substrate/blob/master/frame/vesting/src/lib.rs#L142).

**Comment by wischli** [comment]
@weichweich  @Diiaablo95 The pallet can be reviewed now. Please be as critical as possible, also regarding the documentation. Please have a look at the description of the pallet in the comments below the copyright.

Note: Disabling `locked_transfer` via feature / filtering calls has not been implemented yet. I would like to handle this in another PR. Same holds true for adding `pallet_vesting` directly to our pallets and doing the same thing for `vested_transfer`.

File: pallets/kilt-launch/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:285]: Btw: UnlockingAt will be only filled with 6 accounts.

File: pallets/kilt-launch/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,797 @@
...
+//! - Transfer locked tokens from user-owned account to another. NOTE: This will
+//!   be made available shortly before we remove the sudo key.
+//! - Forcedly (requires sudo) changing the `TransferAccount`.
+//! - Forcedly (requires sudo) removing the KILT balance lock.
+//!
+//! ### Terminology
+//!
+//! - **Lock:** A freeze on a specified amount of an account's free balance
+//!   until a specified block number. Multiple
+//! locks always operate over the same funds, so they "overlay" rather than
```
Comment (Diiaablo95) [inline pallets/kilt-launch/src/lib.rs:49]: Wrong indentation here

File: pallets/kilt-launch/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,797 @@
...
+	// Balance type based on pallet_vesting
+	pub type BalanceOf<T> =
+		<<T as pallet_vesting::Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			// Generate initial custom locking configuration
+			// * who - Account which setting the custom lock for
+			// * length - Number of blocks from  until removal of the lock
```
Comment (Diiaablo95) [inline pallets/kilt-launch/src/lib.rs:204]: from...?

File: pallets/kilt-launch/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,797 @@
...
+						who.to_string()
+					);
+					assert!(
+						!<Vesting<T>>::contains_key(who),
+						"Account with address {:?} must not occur twice in vesting",
+						who.to_string()
+					);
+
+					let length_as_balance = T::BlockNumberToBalance::convert(length);
+					let per_block = locked / length_as_balance.max(sp_runtime::traits::One::one());
```
Comment (Diiaablo95) [inline pallets/kilt-launch/src/lib.rs:256]: Could we use a checked division, just to be sure in case anything changes in the future?

**Review by Diiaablo95 — APPROVED** [review]
So far so good, I would say 😁 I just spotted a couple of minor comment stuff, and a question about a safe division. It might probably not be a super in-depth review, but it definitely helped me a lot 🙂

File: pallets/kilt-launch/src/lib.rs (hunk 13 above)
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:256]: It's definitely safer. The code was actually taken from the vesting pallet.

File: pallets/kilt-launch/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,797 @@
...
+	fn migrate_kilt_balance_lock(
+		source: &T::AccountId,
+		target: &T::AccountId,
+		// Only used for `locked_transfer`, e.g., it is `None` for migration
+		max_amount: Option<<T as pallet_balances::Config>::Balance>,
+	) -> Result<Weight, DispatchError> {
+		if let Some(LockedBalance::<T> {
+			block: unlock_block,
+			amount: source_amount,
+		}) = <BalanceLocks<T>>::get(&source)
```
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs:724]: Maybe we could do something like this?

```rust
    let inner = if let Some(i) = opt {
        i
    } else {
        return Ok(T::DbWeight::get().reads(1));
    };
```

File: pallets/kilt-launch/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,797 @@
...
+	fn migrate_vesting(source: &T::AccountId, target: &T::AccountId) -> Result<Weight, DispatchError> {
+		if let Some(source_vesting) = Vesting::<T>::take(source) {
+			// Check for an already existing vesting schedule for the target account
+			// which would be the case if the claimer requests migration from multiple
+			// source accounts to the same target
+			let vesting = if let Some(VestingInfo {
+				locked,
+				per_block,
+				starting_block,
+			}) = Vesting::<T>::take(&target)
```
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs:668]: Maybe that's more readable? Not sure... 🤷🏻 

```rust
let vesting = Vesting::<T>::take(&target).map_or(Ok(source_vesting), |VestingInfo {
		locked,
		per_block,
		starting_block,
	}| {
		// Should never throw because all source accounts start vesting in genesis block
		ensure!(
			starting_block == source_vesting.starting_block,
			Error::<T>::ConflictingVestingStarts
		);
		Ok(VestingInfo {
			// We can simply sum `locked` and `per_block` because of the above requirement
			locked: locked.saturating_add(source_vesting.locked),
			per_block: per_block.saturating_add(source_vesting.per_block),
			starting_block,
		})
	})?;

**Review by weichweich — CHANGES_REQUESTED** [review]
I have a few questions and some style remarks. Not sure if my suggestions would make it better. Just wanted to propose them.

File: pallets/kilt-launch/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,797 @@
...
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarking;
+pub mod default_weights;
+
+pub const KILT_LAUNCH_ID: LockIdentifier = *b"kiltlnch";
+pub const VESTING_ID: LockIdentifier = *b"vesting ";
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:136]: It is intentional and we have to keep it to not cause trouble with utility from the vesting pallet. It is definitely hacky but the only other solution I see is to use the vesting pallet from a local copy rather than a dependency and then make the id public. Would you rather do it that way?

File: pallets/kilt-launch/src/lib.rs (hunk 13 above)
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:256]: It doesn't make sense to do a checked division because we check for `length.is_zero()` a couple lines earlier and even safely do a max check with One.

Comment (Diiaablo95) [reply pallets/kilt-launch/src/lib.rs:256]: I would not rely on a check performed at the beginning of an if branch, as that might change in future. Furthermore, now that I see it better, I think it would be nicer to have a guard condition replacing the if condition and not indent the whole block within the if branch.
Plus, how is `value.max(1)` protecting against division by zero? Should it be `value.min(1)`?

File: pallets/kilt-launch/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,797 @@
...
+		/// Because this pallet emits events, it depends on the runtime's
+		/// definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Maximum number of claims which can be migrated in a single call.
+		/// Used for weight estimation.
+		///
+		/// Note: Benchmarks will need to be re-run and weights adjusted if this
+		/// changes.
+		// FIXME: Why does this not work?
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:160]: Leftover from `#[pallet::constant]` not working 😅

File: pallets/kilt-launch/src/lib.rs (hunk 13 above)
Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs:256]: > Plus, how is `value.max(1)` protecting against division by zero? Should it be `value.min(1)`?

That is an error that i make sooooooo often!! Whats the max of 0 and 1?

Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:256]: Yes, sorry, the max is not protecting.

Comment (Diiaablo95) [reply pallets/kilt-launch/src/lib.rs:256]: > > Plus, how is `value.max(1)` protecting against division by zero? Should it be `value.min(1)`?
> 
> That is an error that i make sooooooo often!! Whats the maximum of 0 and 1?

Mhhh, I think we need to move to the irrational numbers to answer correctly 🤔

Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:256]: Thanks @weichweich for pointing this out: If `length_as_balance` is zero, then 0.max(1) == 1.

Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs:256]: `max` is the lower bound, `min` is the upper bound. That's probably where the confusion comes from. 😂

Comment (Diiaablo95) [reply pallets/kilt-launch/src/lib.rs:256]: Anyway, it kind of proved my point that mistakes, of any kind, can happen 😁 so unless there's some big overhead between checked vs unchecked divisions, I would prefer to use the checked one anything I can.

Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:256]: Regarding the guard: I

> Anyway, it kind of proved my point that mistakes, of any kind, can happen 😁 so unless there's some big overhead between checked vs unchecked divisions, I would prefer to use the checked one anywhere I can.

It is overhead and since Parity doesn't do it either, I would not as well: https://github.com/paritytech/substrate/blob/master/frame/vesting/src/lib.rs#L170

Also please note that this function is just the Genesisbuilder, so it won't be called more than once in the lifetime of our chain and the whole pallet will be removed before 2022.

Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:256]: Thus, I would leave it as is. The input for this function is a JSON which has one line for each endowed account which might either have
1. All of their balance available
2. All of their balance vested
3. All of their balanced locked

Thus, there are a lot of 0-rows for locking and vesting.

File: pallets/kilt-launch/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,797 @@
...
+use frame_support::{
+	dispatch::DispatchResultWithPostInfo,
+	inherent::Vec,
+	pallet_prelude::*,
+	sp_runtime::traits::{StaticLookup, Zero},
+	storage::types::StorageMap,
+	traits::{Currency, ExistenceRequirement::AllowDeath, Get, LockIdentifier, LockableCurrency, WithdrawReasons},
+	transactional,
+};
+pub use pallet::*;
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:107]: We did not have to use `pub use`. Fixed in b70bd3d8efa15afb38dc8b360454ab6ab7e3a8e2.

File: pallets/kilt-launch/src/lib.rs (hunk 13 above)
Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs:253]: I would still say that this is a bit more readable?

```suggestion
					let per_block = locked.checked_div(length_as_balance).unwrap_or(locked);
```

Comment (wischli) [reply pallets/kilt-launch/src/lib.rs]: Fixed in 6157802e90054e0c5713bf5847c29e924f22226d

Comment (wischli) [reply pallets/kilt-launch/src/lib.rs]: Fixed guard remark of @Diiaablo95 in ad05d17591d103fa175b65dbbbd71a66a072a36e

> I would not rely on a check performed at the beginning of an if branch, as that might change in future. Furthermore, now that I see it better, I think it would be nicer to have a guard condition replacing the if condition and not indent the whole block within the if branch.

File: pallets/kilt-launch/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,797 @@
...
+		///   BalanceLocks
+		/// - Writes: Locks, Balance, Vesting (if any source is vesting),
+		///   BalanceLocks (if aby source is locking), UnlockingAt (if any
+		///   source is locking)
+		/// - Kills (for sources): Locks, Balance, Vesting (if any source is
+		///   vesting), BalanceLocks (if any source is locking)
+		/// # </weight>
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::migrate_multiple_genesis_accounts_vesting(T::MaxClaims::get()).max(<T as pallet::Config>::WeightInfo::migrate_multiple_genesis_accounts_locking(T::MaxClaims::get())))]
+		#[transactional]
+		pub fn migrate_multiple_genesis_accounts(
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs:486]: As discussed: We keep it as is.

File: pallets/kilt-launch/src/lib.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs]: This is the best that worked for me: d35cf33638fbdd99eca025fc03f817093b2e438e

File: pallets/kilt-launch/src/lib.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs]: As discussed, the `ensure!` breaks it kinda. I hope this will suffice: d35cf33638fbdd99eca025fc03f817093b2e438e

File: pallets/kilt-launch/src/lib.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs]: As discussed: We will not add the `pallet_vesting` manually. Our unit tests will show whenever they change the `LockIdentifier` since we rely on functionality from `pallet_vesting` to update and eventually remove the lock.

File: pallets/kilt-launch/src/lib.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/kilt-launch/src/lib.rs]: I actually meant something like this 🙂 : 299101b7b72f92c4182623d9615402e6ebe5ee38

Now that I touched the code, thought that we maybe want to throw an error in case the lock doesn't exists?

[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs]: If we do want to throw, we have to add an input parameter to `migrate_user`. Because when calling this function, it is not assumed that either `vesting` or `locking` exists for the source, it could just migrate the balance.

[the remark this answers is not in the document]
Comment (wischli) [reply pallets/kilt-launch/src/lib.rs]: > I actually meant something like this 🙂 : [299101b](https://github.com/KILTprotocol/mashnet-node/commit/299101b7b72f92c4182623d9615402e6ebe5ee38)

Looks good, thanks!

**Review by weichweich — APPROVED** [review]
LGTM
